Sources/NIOSOCKS/Messages/ClientRequest.swift

```diff
@@ -0,0 +1,143 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+// MARK: - ClientRequest
+
+/// Instructs the SOCKS proxy server of the target host,
+/// and how to connect.
+public struct ClientRequest: Hashable {
+    
+    /// The SOCKS protocol version - we currently only support v5.
+    public var version: UInt8
+    
+    /// How to connect to the host.
+    public var command: Command
+    
+    /// The target host address.
+    public var addressType: AddressType
+    
+    /// The target host port.
+    public var desiredPort: UInt16
+    
+    /// Creates a new `ClientRequest`.
+    /// - parameter command: How to connect to the host.
+    /// - parameter addressType: The target host address.
+    /// - parameter desiredPort: The target host port.
+    public init(command: Command, addressType: AddressType, desiredPort: UInt16) {
+        self.version = 5
+        self.command = command
+        self.addressType = addressType
+        self.desiredPort = desiredPort
+    }
+    
+}
+
+extension ByteBuffer {
+    
+    @discardableResult mutating func writeClientRequest(_ request: ClientRequest) -> Int {
+        self.writeInteger(request.version)
+        self.writeInteger(request.command.rawValue)
+        self.writeInteger(0, as: UInt8.self)
+        let addressSize = self.writeAddressType(request.addressType)
+        self.writeInteger(request.desiredPort)
+        return 1 + 1 + 2 + addressSize + 2
+    }
+    
+}
+
+// MARK: - Command
+
+/// What type of connection the SOCKS server should establish with
+/// the target host.
+public enum Command: UInt8 {
```

Comment:
Relatedly, should this be an `enum` or a `struct` with `rawValue` and some `static let`s?

Reply:
Yep!